core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -369,6 +370,32 @@ class GroupCoordinator(val brokerId: Int,
     }
   }
 
+  /**
+   * try to complete produce, fetch and delete requests if the HW of partition is incremented. Otherwise, we try to complete
+   * only delayed fetch requests.
+   *
+   * Noted that this method may hold multiple group locks so the caller should NOT hold any group lock
+   * in order to avoid deadlock
+   * @param topicPartitions a map contains the partition and a flag indicting whether the HWM has been changed
+   */
+  private[group] def completeDelayedRequests(topicPartitions: Map[TopicPartition, LeaderHwChange]): Unit =
+    topicPartitions.foreach {
+      case (tp, leaderHWIncremented) => leaderHWIncremented match {
+        case LeaderHwIncremented => groupManager.replicaManager.completeDelayedRequests(tp)
+        case _ => groupManager.replicaManager.completeDelayedFetchRequests(tp)
+      }
+    }
+
+  /**
+   * complete the delayed join requests associated to input group keys.
+   *
+   * Noted that delayedJoin itself uses a lock other than the group lock for DelayedOperation.maybeTryComplete() and
```

Comment:
Hmm, it seems that we are now introducing a new potential deadlock. The conflicting paths are the following.

path 1
hold group lock -> joinPurgatory.tryCompleteElseWatch(delayedJoin) -> watchForOperation (now delayedJoin visible through other threads) -> operation.maybeTryComplete() -> hold delayedJoin.lock

path 2
delayedJoin.maybeTryComplete -> hold hold delayedJoin.lock -> tryComplete() -> hold group lock

Reply:
How about removing inner lock (```ReentrantLock```) from ```DelayedJoin``` ? It seems to me ```DelayedJoin``` does not need the inner lock.

Reply:
another approach is that - we introduce an new method "afterTryComplete" to ```DelayedOperation```. the new method is invoked by ```maybeTryComplete``` after lock is released. ```DelayedJoin``` still pass group lock to ```DelayedOperation``` and use "afterTryComplete" to complete delayed requests

Reply:
@chia7712 : Yes, that's a possibility. It adds some complexity to DelayedOperation. Another possibility is to have a special case to complete the delayed requests from groupManager.storeGroup() in GroupCoordinator.onCompleteJoin() in a separate thread.